external-crates/move/move-compiler/src/parser/syntax.rs

```diff
@@ -2104,6 +2124,61 @@ fn parse_field_annot(context: &mut Context) -> Result<(Field, Type), Box<Diagnos
     Ok((f, st))
 }
 
+// Parse a postfix ability declaration:
+//     "has" <Ability> (, <Ability>)+;
+//  Error if:
+//      * Also has prefix ability declaration
+fn parse_postfix_ability_declarations(
+    infix_ability_declaration_loc: Option<Loc>,
+    abilities: &mut Vec<Ability>,
+    context: &mut Context,
+) -> Result<(), Box<Diagnostic>> {
+    let postfix_ability_declaration =
+        context.tokens.peek() == Tok::Identifier && context.tokens.content() == "has";
+    let has_location = current_token_loc(context.tokens);
+
+    if postfix_ability_declaration
+        && context
+            .env
+            .check_feature(&FeatureGate::PostFixAbilities, None, has_location)
+    {
+        context.tokens.advance()?;
+
+        // Only add a diagnostic about prefix xor postfix ability declarations if the feature is
+        // supported. Otherwise we will already have an error that the `has` is not supported in
+        // that position, and the feature check diagnostic as well, so adding this additional error
+        // could be confusing.
+        if let Some(previous_declaration_loc) = infix_ability_declaration_loc {
+            let msg = "Invalid ability declaration. Abilities cannot be declared in \
+                       infix and postfix positions simultaneously";
```

Comment:
Stating infix + postfix sounds a bit overly compilery to me. Maybe something like 
```suggestion
            let msg = "Duplicate ability declaration. Abilities can be declare before or after the field declarations, but not both";
```